lib/eventmachine.rb

```diff
@@ -751,6 +751,28 @@ def EventMachine::attach_io io, watch_mode, handler=nil, *args
     c
   end
 
+  # Attaches a server IO object or file descriptor to the eventloop.
+  # This function behaves just like start_server but allows you to reuse
+  # an already existing file descriptor instead of having EventMachine create
+  # one. The descriptor must be accept()able and will be set as non-blocking.
+  #
+  # Unlike start_server however, the file descriptor is not closed when
+  # EventMachine is released, so you will have to do any cleanups manually.
+  # If +io+ is an IO object then a reference to it will be kept until
+  # EventMachine is released, so that the file descriptor isn't accidentally
+  # closed by the garbage collector.
+  def EventMachine::attach_server io, handler = nil, *args, &block
+    klass = klass_from_handler(Connection, handler, *args)
+    if io.respond_to?(:fileno)
+      fd = defined?(JRuby) ? JRuby.runtime.getDescriptorByFileno(io.fileno).getChannel : io.fileno
+    else
+      fd = io
+    end
+    s = attach_server_fd(fd)
```

Comment:
Although this was superseded by #465, I am looking into JRuby test failures in #551 and I don't see where attach_server_fd is defined for the JRuby implementation. So even though you're making sure to use `JRuby.runtime.getDescriptorByFileno`, it's not getting passed into a valid function, AFAICT.